### ISSUE TITLE
Request for JSON Header Dictionary CleanUp

### ISSUE THREAD

**Issue by fjaconete** [rebus-org/Rebus#701]
I'm facing a hard time figuring out what is going on with Rebus MSMQ headers.
According to [this track on StackOverflow](https://stackoverflow.com/questions/50299534/rebus-puting-weird-characters-at-header-dictionary-json-written-at-msmq-extensio) the MSMQ message is receiving some weird characters at the end of the JSON, causing problems for Rebus to deserialize it.
What could be included at Rebus.Serialization implementation is kinda a Hygiene for the received json text, before trying to deserialize it.

Since the ThreadPoolWorker encapsulate the SerializationException thrown by Newtonsoft, I'm also not able to create any kind of error handling behavior either.

At Rebus.HeaderSerializer.Deserialize it could have some code to clean up possible harmfull JSON before deserialize it, avoiding some binary tampering on Extension property at MSMQ.

Here is a suggestion for implementation:

- Changed Deserialize method
`public Dictionary<string, string> Deserialize(byte[] bytes)
        {
            var jsonString = HigineJson(Encoding.GetString(bytes));

            return DeserializeFromString(jsonString);
        }`

- Support methods:
`private string HigineJson(string sourceExtText)
        {
            var items = sourceExtText.Split(new[] { '}' }, StringSplitOptions.RemoveEmptyEntries);
            var idx = 0;
            var result = items[idx] + "}";

            while (!IsValidJson(result) && idx < items.Length - 1)
            {
                idx++;
                result += items[idx] + "}";
            }

            if (!IsValidJson(result))
                throw new SerializationException($"Invalid JSON format '{result}'");

            return result;
        }

        private bool IsValidJson(string strInput)
        {
            strInput = strInput.Trim();
            if ((strInput.StartsWith("{") && strInput.EndsWith("}")) || (strInput.StartsWith("[") && strInput.EndsWith("]")))
            {
                try
                {
                    var obj = JToken.Parse(strInput);
                    return true;
                }
                catch (JsonReaderException jex)
                {
                    Console.WriteLine(jex.Message);
                    return false;
                }
                catch (Exception ex) //some other exception
                {
                    Console.WriteLine(ex.ToString());
                    return false;
                }
            }
            return false;
        }`

I'm considering create a fork and recompile Rebus in a local version in order to reference it on my project, however it could a great asset to have it embedded at Rebus default implementation.

**Comment by mookid8000** [comment]
So sorry to learn that you are still fighting with that crazy MSMQ extension issue 😩 

However, I am hesitant to change Rebus core because of this – I've never experienced it myself(*), and I haven't heard of anyone with similar issues.

If you can reproduce the issue with Rebus on your own machine, I would be happy to debug the code and figure out what causes it.

But I still do suspect that this garbage comes from something in your hosting environment.

Could you maybe once more check, if something like [dynatrace](https://www.dynatrace.com/support/doc/appmon/application-monitoring/agents/net-agent/msmq/) is running? See [here](https://answers.dynatrace.com/questions/92449/msmq-messages-being-sent-to-error-queue.html) for an example on someone, whose MSMQ messages are being mangled by it.



----
(*) or rather, I have – but [as I told you](https://stackoverflow.com/a/50307289/6560), it turned out to be caused by something out of our hands

**Comment by fjaconete** [comment]
Unfortunately there is nothing going on over MSMQ except for my software. We did check about that.
What we are doing now is a full detailement of the hosting environment. All about installed softwares, anti-virus, firewall, network filters and so on. I'm with you in that, my understanding is that something at the environment is messing up with JSON serialization in time of writing MSMQ message.

However what is giving me a rough time about Rebus is that there is no support for handling errors in that part of Rebus implementation, so Rebus is catching Serialization Exception, writing logs and just that, I can't do anything about it.
If there was an abstraction in wich I could inject an error handling implementation of mine, I could fix the JSON message (just like I did in a external windows service, switching queues) and then enqueue the message back for good processing.
I've try using Rebus 4.2.1 and provide a JsonSerializerSettings with a Error delegate pointing to a local method, but this error at deserialization doesn't fire this event as well.

Can you help me out with that!?
Until I figure it out what is wrong with the server that is causing this issue, I need to put running some workaround for the customer production not get affected anymore.

Thanks a lot.

**Comment by mookid8000** [comment]
> (...) something at the environment is messing up with JSON serialization in time of writing MSMQ message

That's exactly my point – I do NOT think something is wrong with how the data is serialized (or deserialized for that matter) – I think the message gets corrupted some time between sending and receiving it again.

I have an idea on how you can solve this problem for yourself, though – simply go and copy `MsmqTransport` from [this repository](https://github.com/rebus-org/Rebus.Msmq) and bring along everything it depends on.

When you have the code under your control, you can introduce the fix you proposed.

> (...) there is no support for handling errors in that part of Rebus implementation

And yes, you are right that Rebus does not provide a way to rectify the situation when an incoming message has been corrupted by the transport. It would have, if I thought that this problem was commonplace, but unfortunately I think you've been hit by a very niche type of error.

I understand if you think I am being unyielding here, but my goal is to keep the code base clean and focused, and in doing that I cannot let it be distracted by too many niche problems like this, however despairing it might be to you.

> I've try using Rebus 4.2.1 and provide a JsonSerializerSettings with a Error delegate pointing to a local method, but this error at deserialization doesn't fire this event as well.

Yeah, and this is simply because that only affects how Rebus serializes/deserializes the message body. How the `Extension` is serialized is (and should never by) of any concern to the user, because it's just a bunch of key/value pairs, and usually that part is taken case of by a built-in mechanism in the transport – MSMQ is just whack in that regard, because there is no built-in header mechanism, and therefore the transport needs to do it like that.

**TL;DR**: Copy the code from [Rebus' MSMQ transport](https://github.com/rebus-org/Rebus.Msmq) and introduce the fix you proposed. I would be a little bit worried though, that – if it's not some kind of MSMQ surveillance that is messing with the messages – something is really wrong, and message bodies could be corrupted too. That would suck.

Sorry, but your situation sucks. I hope you can make it work 😃
